### PR TITLE
Floral spines [New Plant Gene]

### PR DESCRIPTION
**Pull request by DrDuckedGoose**
## About The Pull Request

**Floral Spines** is a new gene for plants that causes the plant to stick to players like a throwing star, when thrown. Additionally, adding hypodermic prickles to a plant with this trait also makes the impact deal damage, using the formula ``(seed.potency/4.7)`` You can also combine the slippery skin trait to allow the plant to embed when slipped on. 

**Floral Spines** naturally occur in lavaland Cactus, and Starthistle.

## Why It's Good For The Game

This PR not only allows botanists to grow and engineer unique tools but, also offers the potential for further development and use of the core concept. 

## Changelog
:cl:
add: Adds Floral Spines
tweak: Lavaland Cactus now has Floral Spines
tweak: Starthistle now has Floral Spines
/:cl:

### CODE REVIEW

File: austation/code/modules/hydroponics/grown.dm [hunk 1]
```diff
@@ -1,3 +1,18 @@
...
+
+/obj/item/reagent_containers/food/snacks/grown/Initialize(mapload, obj/item/seeds/new_seed)
+	. = ..()
+
+	if(seed.get_gene(/datum/plant_gene/trait/spines))
+		embedding = EMBED_HARMLESS
+		embedding["embed_chance"] = 300 //300 is better than 100 ;)
+		updateEmbedding()
+	
+	grown_overlay = mutable_appearance(icon, icon_state)
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/grown.dm:14]: what's the reason you need a dedicated variable to store this? why not just generate it at the time of embed?

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 2]
```diff
@@ -34,3 +34,38 @@
...
+				G.target = target
+
+				G.grown_overlay.layer = FLOAT_LAYER
+				G.target.add_overlay(G.grown_overlay, TRUE)
+			
+				var/P = G.seed.get_gene(/datum/plant_gene/trait/stinging)
+				var/mob/living/L = target
+				L.adjustBruteLoss((G.seed.potency/4.7)*P)
+
+				G.tryEmbed(target, TRUE, TRUE)
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:55]: condense these
```suggestion
	if(G.seed && G.seed.get_gene(/datum/plant_gene/trait/spines) && istype(target, /mob/living))
		G.target = target

		G.grown_overlay.layer = FLOAT_LAYER
		G.target.add_overlay(G.grown_overlay, TRUE)
			
		var/P = G.seed.get_gene(/datum/plant_gene/trait/stinging)
		var/mob/living/L = target
		L.adjustBruteLoss((G.seed.potency/4.7)*P)

		G.tryEmbed(target, TRUE, TRUE)
```

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 3]
```diff
@@ -34,3 +34,38 @@
...
+					G.target = target
+
+					G.grown_overlay.layer = FLOAT_LAYER
+					G.target.add_overlay(G.grown_overlay, TRUE)
+				
+					var/P = G.seed.get_gene(/datum/plant_gene/trait/stinging)
+					var/mob/living/L = target
+					L.adjustBruteLoss((G.seed.potency/4.7)*P)
+
+					G.tryEmbed(target, TRUE, TRUE)
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:71]: holy shit dude
```suggestion
	if(!..() && G.seed && G.seed.get_gene(/datum/plant_gene/trait/spines) && istype(target, /mob/living))
		G.target = target

		G.grown_overlay.layer = FLOAT_LAYER
		G.target.add_overlay(G.grown_overlay, TRUE)
				
		var/P = G.seed.get_gene(/datum/plant_gene/trait/stinging)
		var/mob/living/L = target
		L.adjustBruteLoss((G.seed.potency/4.7)*P)

		G.tryEmbed(target, TRUE, TRUE)
```

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 4]
```diff
@@ -34,3 +34,38 @@
...
+			if(G.seed.get_gene(/datum/plant_gene/trait/spines))
+				if(istype(target, /mob/living))
+					G.target = target
+
+					G.grown_overlay.layer = FLOAT_LAYER
+					G.target.add_overlay(G.grown_overlay, TRUE)
+				
+					var/P = G.seed.get_gene(/datum/plant_gene/trait/stinging)
+					var/mob/living/L = target
+					L.adjustBruteLoss((G.seed.potency/4.7)*P)
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:69]: this is going to runtime if the plant doesn't have the stinging gene, as P will be null

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 5]
```diff
@@ -34,3 +34,38 @@
...
+				L.adjustBruteLoss((G.seed.potency/4.7)*P)
+
+				G.tryEmbed(target, TRUE, TRUE)
+
+/datum/plant_gene/trait/spines/on_throw_impact(obj/item/reagent_containers/food/snacks/grown/G, atom/target)
+	if(!..())
+		if(G.seed)
+			if(G.seed.get_gene(/datum/plant_gene/trait/spines))
+				if(istype(target, /mob/living))
+					G.target = target
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:62]: this variable should be set only if the embedding actually succeeds

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 6]
```diff
@@ -34,3 +34,38 @@
...
+
+//Too close to Neck-Code TM
+/datum/plant_gene/trait/spines
+	name = "Floral Spines"
+
+/datum/plant_gene/trait/spines/on_slip(obj/item/reagent_containers/food/snacks/grown/G, atom/target)
+	if(G.seed)
+		if(G.seed.get_gene(/datum/plant_gene/trait/spines))
+			if(istype(target, /mob/living))
+				G.target = target
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:46]: this variable should be set only if the embedding actually succeeds

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 7]
```diff
@@ -34,3 +34,38 @@
...
+
+				G.grown_overlay.layer = FLOAT_LAYER
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:48]: why not set this when you first create the overlay in Initialize()?

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 8]
```diff
@@ -34,3 +34,38 @@
...
+				G.target.add_overlay(G.grown_overlay, TRUE)
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:49]: this should only be run if the embed actually succeeds

File: code/modules/hydroponics/grown/misc.dm [hunk 9]
```diff
@@ -13,7 +13,7 @@
 	potency = 10
 	growthstages = 3
 	growing_icon = 'icons/obj/hydroponics/growing_flowers.dmi'
-	genes = list(/datum/plant_gene/trait/plant_type/weed_hardy)
+	genes = list(/datum/plant_gene/trait/plant_type/weed_hardy, /datum/plant_gene/trait/spines)
```
Comment (MCterra10) [inline code/modules/hydroponics/grown/misc.dm:16]: please comment all unmodular code changes, thanks
```suggestion
	genes = list(/datum/plant_gene/trait/plant_type/weed_hardy, /datum/plant_gene/trait/spines) // austation -- adds floral spines gene
```

File: code/modules/mining/lavaland/ash_flora.dm [hunk 10]
```diff
@@ -235,7 +235,7 @@
 	species = "cactus"
 	plantname = "Fruiting Cactus"
 	product = /obj/item/reagent_containers/food/snacks/grown/ash_flora/cactus_fruit
-	genes = list(/datum/plant_gene/trait/fire_resistance)
+	genes = list(/datum/plant_gene/trait/fire_resistance, /datum/plant_gene/trait/spines)
```
Comment (MCterra10) [inline code/modules/mining/lavaland/ash_flora.dm:238]: same here
```suggestion
	genes = list(/datum/plant_gene/trait/fire_resistance, /datum/plant_gene/trait/spines) // austation -- adds floral spines gene
```

**Review by MCterra10 — CHANGES_REQUESTED** [review]
please remove all changes to code/modules/grown.dm
the only changes there are you fucking around with newlines, which is going to cause code conflicts

File: austation/code/modules/hydroponics/grown.dm [hunk 11]
```diff
@@ -1,3 +1,19 @@
+/obj/item/reagent_containers/food/snacks/grown
+	var/mutable_appearance/grown_overlay //Used for 'Floral Spines' plant gene
+	var/mob/living/target
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/grown.dm:3]: these vars could all be moved to the plant gene itself, rather than being on the plants

**Review by MCterra10 — CHANGES_REQUESTED** [review]
Some of the reviews were not resolved properly. Please deal with those or give some reason for not addressing them.

In addition, please add/modify the Destroy() proc for plants to remove the overlay for their target, if the plant gets deleted. else the overlay would be permanent in that case.

File: austation/code/modules/hydroponics/grown.dm (hunk 11 above)
Comment (DrDuckedGoose) [reply austation/code/modules/hydroponics/grown.dm:3]: It's easier in terms of initializing.

File: austation/code/modules/hydroponics/grown.dm [hunk 12]
```diff
@@ -32,3 +44,13 @@
 		reagents.reaction(A)
 
 	qdel(src)
+
+/obj/item/reagent_containers/food/snacks/grown/unembedded()
+	. = ..()
+	for(var/datum/plant_gene/trait/spines/S in seed.genes)
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/grown.dm:50]: can there be more than one??
applies to the other for loops as well
surely you can just use get_gene() like you do everywhere else

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 13]
```diff
@@ -34,3 +34,28 @@
...
 	G.reagents.handle_reactions()
+
+//Too close to Neck-Code TM
+/datum/plant_gene/trait/spines
+	name = "Floral Spines"
+	var/mob/victim
+	var/mutable_appearance/grown_overlay
+
+/datum/plant_gene/trait/spines/on_slip(obj/item/reagent_containers/food/snacks/grown/G, atom/target)
+	if(G.seed && G.seed.get_gene(/datum/plant_gene/trait/spines) && istype(target, /mob/living))
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:45]:
```suggestion
	if(G?.seed.get_gene(/datum/plant_gene/trait/spines) && istype(target, /mob/living))	
```

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 14]
```diff
@@ -34,3 +34,28 @@
...
+/datum/plant_gene/trait/spines
+	name = "Floral Spines"
+	var/mob/victim
+	var/mutable_appearance/grown_overlay
+
+/datum/plant_gene/trait/spines/on_slip(obj/item/reagent_containers/food/snacks/grown/G, atom/target)
+	if(G.seed && G.seed.get_gene(/datum/plant_gene/trait/spines) && istype(target, /mob/living))	
+		var/P = G.seed.get_gene(/datum/plant_gene/trait/stinging)
+		var/mob/living/L = target
+		if(P) L.adjustBruteLoss((G.seed.potency/4.7)*P)
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:48]: :vomiting_face: 
```suggestion
		if(P)
			L.adjustBruteLoss((G.seed.potency/4.7)*P)
```

File: austation/code/modules/hydroponics/plant_genes.dm [hunk 15]
```diff
@@ -34,3 +34,28 @@
...
+
+		if(G.tryEmbed(target, TRUE, TRUE))
+			target.add_overlay(grown_overlay, TRUE)
+
+/datum/plant_gene/trait/spines/on_throw_impact(obj/item/reagent_containers/food/snacks/grown/G, atom/target)
+	if(!..() && G?.seed.get_gene(/datum/plant_gene/trait/spines) && istype(target, /mob/living))		
+		var/P = G.seed.get_gene(/datum/plant_gene/trait/stinging)
+		var/mob/living/L = target
+		if(P) L.adjustBruteLoss((G.seed.potency/4.7)*P)
```
Comment (MCterra10) [inline austation/code/modules/hydroponics/plant_genes.dm:57]: :vomiting_face: 
```suggestion
		if(P)
			L.adjustBruteLoss((G.seed.potency/4.7)*P)
```

File: austation/code/modules/hydroponics/grown.dm [hunk 16]
```diff
@@ -32,3 +44,13 @@
...
+		if(S.victim)S.victim.cut_overlay(S.grown_overlay, TRUE)
```
[the remark this answers is not in the document]
Comment (DrDuckedGoose) [reply austation/code/modules/hydroponics/grown.dm:51]: https://www.youtube.com/watch?v=oL2B-AAnsHo

File: austation/code/modules/hydroponics/grown.dm [hunk 17]
```diff
@@ -1,3 +1,15 @@
+/obj/item/reagent_containers/food/snacks/grown/Initialize(mapload, obj/item/seeds/new_seed)
+	. = ..()
+
+	if(seed.get_gene(/datum/plant_gene/trait/spines))
+		embedding = EMBED_HARMLESS
+		embedding["embed_chance"] = 300 //300 is better than 100 ;)
+		updateEmbedding()
+
+		for(var/datum/plant_gene/trait/spines/S in seed.genes)
```
[the remark this answers is not in the document]
Comment (DrDuckedGoose) [reply austation/code/modules/hydroponics/grown.dm]: Yeah, not sure why i didn't consider that. Honestly surprised how I looked over that stuff.

**Review by MCterra10 — APPROVED** [review]

**Pull request merged**